src/shm/pool/multi.rs

```diff
@@ -0,0 +1,295 @@
+//!             and trigger a redraw when it fires.
+//!             (not shown in this example)
+//!         */
+//!     }
+//! }
+//!
+//! fn draw(slice: &mut [u8]) {
+//!     todo!()
+//! }
+//!
+//! ```
+//!
+
+use std::io;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use wayland_client::{
+    protocol::{wl_buffer, wl_shm, wl_shm_pool},
+    ConnectionHandle, Proxy, WEnum,
+};
+
+use super::raw::RawPool;
+
+/// This pool manages buffers associated with keys.
+/// Only one buffer can be attributed to a given key.
+#[derive(Debug)]
+pub struct MultiPool<K: PartialEq + Clone> {
+    buffer_list: Vec<BufferHandle<K>>,
+    pub(crate) inner: RawPool,
+}
+
+#[derive(Debug)]
+struct BufferHandle<K: PartialEq + Clone> {
+    free: Arc<AtomicBool>,
+    size: usize,
+    used: usize,
+    offset: usize,
+    buffer: Option<wl_buffer::WlBuffer>,
+    key: K,
+}
+
+impl<E: PartialEq + Clone> From<RawPool> for MultiPool<E> {
+    fn from(inner: RawPool) -> Self {
+        Self { buffer_list: Vec::new(), inner }
+    }
+}
+
+impl<K: PartialEq + Clone> MultiPool<K> {
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, conn: &mut ConnectionHandle) -> io::Result<()> {
+        self.inner.resize(size, conn)
+    }
+    /// Removes the buffer with the given key from the pool and rearranges the others
+    pub fn remove(&mut self, key: &K, conn: &mut ConnectionHandle) {
+        if let Some((i, buffer)) = self.buffer_list.iter().enumerate().find(|b| b.1.key.eq(key)) {
+            let mut offset = buffer.offset;
+            self.buffer_list.remove(i);
+            for buffer_handle in &mut self.buffer_list {
+                if buffer_handle.offset > offset && buffer_handle.free.load(Ordering::Relaxed) {
+                    if let Some(buffer) = buffer_handle.buffer.take() {
+                        buffer.destroy(conn);
+                    }
+                    std::mem::swap(&mut buffer_handle.offset, &mut offset);
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+    /// Returns the buffer associated with the given key and its offset (usize) in the mempool.
+    ///
+    /// The offset can be used to determine whether or not a buffer was moved in the mempool
+    /// and by consequence if it should be damaged partially or fully.
+    ///
+    /// When it's not possible to use the buffer associated with the key, None is returned.
+    pub fn create_buffer(
+        &mut self,
+        width: i32,
+        stride: i32,
+        height: i32,
+        key: &K,
+        format: wl_shm::Format,
+        conn: &mut ConnectionHandle,
+    ) -> Option<(usize, wl_buffer::WlBuffer, &mut [u8])>
```

Comment:
I feel it would be more ergonomic to split this function in two.

The first function would have a signature like this one to create the buffer, associating a key with the buffer slot.

The second function would act like an `acquire` function of sorts, returning a reference to the buffer and the slice representing it's content.

Of course this splitting would mean we need a way to resize a specific buffer in there based on it's key if it isn't in use.

Reply:
I'm not sure I get your point. Could you explain more how it would improve the ergonomics, and how it'd be used?

Reply:
One way to look at a multi pool is a sort of `HashMap<K, WlBuffer> that guards lookup of an entry if the server has not released the buffer.

Essentially my idea above was to make the semanatics of the equivalentish insert, get, remove, etc operations of the multi pool more explicit to a user of the api.

Currently the smenatics of get, insert and resize all correspond to the `create_buffer` function.